src/mlpack/core/data/tokenizers/BERT_tokenizer_impl.hpp

```diff
@@ -0,0 +1,276 @@
+    return false;
+}
+
+BasicTokenizer::BasicTokenizer(bool doLowerCase=true) 
+    : mDoLowerCase(doLowerCase) {
+}
+
+std::vector<std::wstring> BasicTokenizer::runSplitOnPunc(const std::wstring& text) const {
+    size_t i = 0;
+    bool startNewWord = true;
+    std::vector<std::wstring> output;
+    while (i < text.size()) {
+        wchar_t ch = text[i];
+        if (isPunctuation(ch)) {
+            output.push_back(std::wstring(&ch, 1));
+            startNewWord = true;
+        }
+        else {
+            if (startNewWord) output.push_back(std::wstring());
+            startNewWord = false;
+            output[output.size() - 1] += ch;
+        }
+        i++;
+    }
+    return output;
+}
+
+std::vector<std::wstring> BasicTokenizer::tokenize(const std::string& text) const {
+    std::wstring nText;
+    convertToUnicode(text, nText);
+    nText = cleanText(nText);
+
+    const std::vector<std::wstring>& origTokens = whitespaceTokenize(nText);
+    std::vector<std::wstring> splitTokens;
+    bool is_bert_tkn;
+    for (std::wstring token : origTokens) {
+        is_bert_tkn = false;
+        //If is_bert_tkn is true, then we won't convert the token to smaller case.
+        //The following list contains the list of BERT symbols.
+        std::wstring predefined_bert_tkns[3] = {L"[CLS]", L"[SEP]", L"[MASK]"};
+        for(int i=0; i<3;i++){
+            if(token.find(predefined_bert_tkns[i]) != std::string::npos){
+                is_bert_tkn = true;
+            }
+        }
+
+        if (mDoLowerCase and !is_bert_tkn) {
+            //Temporary variable used for converting string to lower.
+            std::string token_s;
+            //Convert to std::string format.
+            convertFromUnicode(token, token_s);
+            //Converting to lower case.
+            transform(token_s.begin(), token_s.end(), token_s.begin(), ::tolower);
+            //Convert to std::wstring format.
+            convertToUnicode(token_s, token);
+        }
+        const auto& tokens = runSplitOnPunc(token);
+        splitTokens.insert(splitTokens.end(), tokens.begin(), tokens.end());
+    }
+
+    //Ensuring that BERT symbols remain intact after Basic processing.
+    std::wstring processed_text = str_replacer(boost::join(splitTokens, L" "), L"[ CLS ]", L"[CLS]");
+    processed_text = str_replacer(processed_text, L"[ MASK ]", L"[MASK]");
+    processed_text = str_replacer(processed_text, L"[ SEP ]", L"[SEP]");
+
+    return whitespaceTokenize(processed_text);
+}
+
+WordpieceTokenizer::WordpieceTokenizer(const std::shared_ptr<Vocab> vocab, const std::wstring& unkToken, size_t maxInputCharsPerWord)
+    : mVocab(vocab),
+    mUnkToken(unkToken),
+    mMaxInputCharsPerWord(maxInputCharsPerWord) {
+}
+
+std::vector<std::wstring> WordpieceTokenizer::tokenize(const std::wstring& text) const {
+    std::vector<std::wstring> outputTokens;
+    for (auto& token : whitespaceTokenize(text)) {
```

Comment:
I'd prefer to avoid `auto` if possible since the type here is simple.  It makes the code harder to understand, since I can't immediately see what type `whitespaceTokenize()` returns.